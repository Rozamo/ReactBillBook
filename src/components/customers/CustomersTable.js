import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import loadingGif from '../../images/loading.gif';
import LoadData from '../helper/api/LoadData';
import BlueButton from '../bluebutton/BlueButton';
import Table from '../helper/table/Table';

export default function CustomersTable(props) {
  const [error, setError] = useState(null);
  const [isLoaded, setIsLoaded] = useState(false);
  const [items, setItems] = useState([]);

  async function getData() {
    const data = await LoadData('customers');

    if (data && data.entity === 'collection') {
      setIsLoaded(true);
      setItems(data.items);
    } else if (data) {
      setError(data);
      setIsLoaded(true);
    }
  }

  useEffect(() => {
    getData();
  }, []);

  function handleContent() {
    if (error) {
      return (
        <div>
          Error:
          {error.message}
        </div>
      );
    }
    if (!isLoaded) return <img src={loadingGif} alt="Loading...." id="load-img" />;

    const submitSuccess = props.location.state ? props.location.state.submitSuccess : false;
    return Table(items, ['NAME', 'PHONE', 'EMAIL', 'CREATED ON'], ['name', 'contact', 'email', 'created_at'], submitSuccess);
  }

  return (
    <div className="content">
      <div className="top-panel">
        <h1 id="title">Customers</h1>
        <Link to="/customers/create">
          <BlueButton sidebarChoice="customers" contentChoice="list" />
        </Link>
      </div>
      {handleContent()}
    </div>
  );
}

CustomersTable.propTypes = {
  location: PropTypes.object.isRequired,
};
