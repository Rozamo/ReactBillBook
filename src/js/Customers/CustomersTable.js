import React, { useState, useEffect } from 'react';
import {Link} from 'react-router-dom';
import load_gif from '../../images/load.gif';
import LoadData from '../Helper/API/LoadData';
import TableHead from '../Helper/TableMaker/TableHead';
import TableBody from '../Helper/TableMaker/TableBody';
import BlueButton from '../BlueButton/BlueButton';

export default function ItemsTable() {
    const [error, setError] = useState(null);
    const [isLoaded, setIsLoaded] = useState(false);
    const [items, setItems] = useState([]);
    async function getData() {
        const data = await LoadData("customers");
        if (data && data.entity === 'collection') {
            setIsLoaded(true);
            setItems(data.items);
        }
        else if (data) {
            setError(data);
            setIsLoaded(true);
        }
    }
    useEffect(() => {
        getData();
    });
    function handleContent() {
        if (error)
            return <div>Error: {error.message}</div>;
        else if (!isLoaded)
            return <img src={load_gif} alt="Loading...." id="load-img"></img>;
        else return <table className="inv-table" id="inv-table">
            {TableHead('NAME', 'PHONE', 'EMAIL', 'CREATED ON')}
            {TableBody(items, 'name', 'contact', 'email', 'created_at')}
        </table>;
    }
    return <div className="content">
        <div className="top-panel">
            <h1 id="title">Customers</h1>
            <Link to='/customers/create'>
                <BlueButton sidebar_choice="customers" content_choice="list"/>
            </Link>
        </div>     
        {handleContent()}        
    </div>;
}
