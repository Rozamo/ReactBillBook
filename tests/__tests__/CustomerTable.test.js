import React from "react";
// import { unmountComponentAtNode, render } from "react-dom";
import { act } from "react-dom/test-utils";
import { BrowserRouter as Router } from "react-router-dom";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { rest } from "msw";
import { setupServer } from "msw/node";
import "@testing-library/jest-dom/extend-expect";
// import "isomorphic-fetch";

import CustomersTable from "../../src/components/customers/CustomersTable";

// global.fetch = fetch;

const fakeData = {
  entity: "collection",
  items: [
    {
      name: "Jmp",
      email: "J@rzp.com",
      age: "32",
      address: "123, Charming Avenue",
      amount: 1231,
    },
  ],
};

// const server = setupServer(
//   rest.get(
//     "https://rzp-training.herokuapp.com/team1/customers",
//     (req, res, ctx) => {
//       return res(ctx.json(fakeData));
//     }
//   )
// );

// beforeAll(() => server.listen());
// afterEach(() => server.resetHandlers());
// afterAll(() => server.close());

test("renders CustomersTable", async () => {
  global.fetch = jest.fn().mockImplementation(() =>
    Promise.resolve({
      json: () => new Promise((resolve) => resolve(fakeData)),
    })
  );

  const { container } = render(
    <Router>
      <CustomersTable location={{ state: { submitSuccess: true } }} />
    </Router>
  );

  expect(container.querySelector("#load-img").getAttribute("alt")).toBe(
    "Loading...."
  );

  expect(container.querySelector(".content")).toMatchSnapshot();

  await waitFor(() => screen);

  expect(container.querySelector(".content")).toMatchSnapshot();

  expect(fetch).toHaveBeenCalledTimes(1);

  global.fetch.mockClear();
  delete global.fetch;
});

test("handles error CustomersTable", async () => {
  global.fetch = jest.fn().mockImplementation(() =>
    Promise.resolve({
      ok: false,
      json: () => new Promise((resolve) => resolve({ status: 500 })),
    })
  );
  //   server.use(
  //     rest.get("", (req, res, ctx) => {
  //       return res(ctx.status(500));
  //     })
  //   );

  const { container } = render(
    <Router>
      <CustomersTable location={{ state: { submitSuccess: true } }} />
    </Router>
  );
  expect(container.querySelector("#load-img").getAttribute("alt")).toBe(
    "Loading...."
  );

  expect(container).toMatchSnapshot();

  await waitFor(() => screen);

  expect(container.querySelector(".content")).toMatchSnapshot();

  global.fetch.mockClear();
  delete global.fetch;
});

// jest.useFakeTimers();

// let container = null;
// beforeEach(() => {
//   // setup a DOM element as a render target
//   container = document.createElement("div");
//   document.body.appendChild(container);
// });

// afterEach(() => {
//   // cleanup on exiting
//   unmountComponentAtNode(container);
//   container.remove();
//   container = null;
// });

// describe("CustomersTable test", () => {
//   it("should render cust table", async () => {
//     const fakeData = {
//       entity: "collection",
//       items: [
//         {
//           name: "Jmp",
//           email: "J@rzp.com",
//           age: "32",
//           address: "123, Charming Avenue",
//           amount: 1231,
//         },
//       ],
//     };
//     global.fetch = jest.fn().mockImplementation(() =>
//       Promise.resolve({
//         json: () =>
//           new Promise((resolve) => setTimeout(() => resolve(fakeData), 1000)),
//       })
//     );

//     await act(async () => {
//       render(
//         <Router>
//           <CustomersTable location={{ state: { submitSuccess: true } }} />
//         </Router>,
//         container
//       );
//     });
//     expect(container.textContent).toContain("Customers");
//     expect(container.innerHTML).toMatchInlineSnapshot(
//       `"<div class=\\"content\\"><div class=\\"top-panel\\"><h1 id=\\"title\\">Customers</h1><a href=\\"/customers/create\\"><button type=\\"button\\" id=\\"button\\" value=\\"\\"><img src=\\"test-file-stub\\" id=\\"floppy\\" alt=\\"New\\">New Customer</button></a></div><img src=\\"test-file-stub\\" alt=\\"Loading....\\" id=\\"load-img\\"></div>"`
//     );
//     // jest.runOnlyPendingTimers();

//     // act(() => {
//     //   jest.advanceTimersByTime(4000);
//     // });

//     await container;

//     expect(container.innerHTML).toMatchInlineSnapshot(
//       `"<div class=\\"content\\"><div class=\\"top-panel\\"><h1 id=\\"title\\">Customers</h1><a href=\\"/customers/create\\"><button type=\\"button\\" id=\\"button\\" value=\\"\\"><img src=\\"test-file-stub\\" id=\\"floppy\\" alt=\\"New\\">New Customer</button></a></div><img src=\\"test-file-stub\\" alt=\\"Loading....\\" id=\\"load-img\\"></div>"`
//     );

//     const newCustBtn = document.querySelector("button");
//     expect(newCustBtn.textContent).toContain("New Customer");

//     act(() => {
//       newCustBtn.dispatchEvent(new MouseEvent("click", { bubbles: true }));
//     });

//     expect(container.innerHTML).toMatchInlineSnapshot(
//       `"<div class=\\"content\\"><div class=\\"top-panel\\"><h1 id=\\"title\\">Customers</h1><a href=\\"/customers/create\\"><button type=\\"button\\" id=\\"button\\" value=\\"\\"><img src=\\"test-file-stub\\" id=\\"floppy\\" alt=\\"New\\">New Customer</button></a></div><img src=\\"test-file-stub\\" alt=\\"Loading....\\" id=\\"load-img\\"></div>"`
//     );

//     global.fetch.mockClear();
//     delete global.fetch;
//   });
// });

// import "isomorphic-fetch";
// // global.fetch = fetch;

// test("real fetch call CustomersTable", async () => {
//   global.fetch = fetch;

//   const res = await fetch("https://jsonplaceholder.typicode.com/users/1");
//   const result = await res.json();
//   expect(result.name).toBe("Leanne Graham"); // Success!

//   await act(async () => {
//     render(
//       <Router>
//         <CustomersTable location={{ state: { submitSuccess: true } }} />
//       </Router>,
//       container
//     );

//     expect(container.innerHTML).toMatchInlineSnapshot(
//       `"<div class=\\"content\\"><div class=\\"top-panel\\"><h1 id=\\"title\\">Customers</h1><a href=\\"/customers/create\\"><button type=\\"button\\" id=\\"button\\" value=\\"\\"><img src=\\"test-file-stub\\" id=\\"floppy\\" alt=\\"New\\">New Customer</button></a></div><img src=\\"test-file-stub\\" alt=\\"Loading....\\" id=\\"load-img\\"></div>"`
//     );

//     // act(() => {
//     //   jest.advanceTimersByTime(4000);
//     // });

//     expect(container.innerHTML).toMatchInlineSnapshot(
//       `"<div class=\\"content\\"><div class=\\"top-panel\\"><h1 id=\\"title\\">Customers</h1><a href=\\"/customers/create\\"><button type=\\"button\\" id=\\"button\\" value=\\"\\"><img src=\\"test-file-stub\\" id=\\"floppy\\" alt=\\"New\\">New Customer</button></a></div><img src=\\"test-file-stub\\" alt=\\"Loading....\\" id=\\"load-img\\"></div>"`
//     );

//     global.fetch.mockClear();
//     delete global.fetch;
//   });
// });
