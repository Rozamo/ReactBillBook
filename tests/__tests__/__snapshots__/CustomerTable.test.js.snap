// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles error CustomersTable 1`] = `
<div>
  <div
    class="content"
  >
    <div
      class="top-panel"
    >
      <h1
        id="title"
      >
        Customers
      </h1>
      <a
        href="/customers/create"
      >
        <button
          id="button"
          type="button"
          value=""
        >
          <img
            alt="New"
            id="floppy"
            src="test-file-stub"
          />
          New Customer
        </button>
      </a>
    </div>
    <img
      alt="Loading...."
      id="load-img"
      src="test-file-stub"
    />
  </div>
</div>
`;

exports[`handles error CustomersTable 2`] = `
<div
  class="content"
>
  <div
    class="top-panel"
  >
    <h1
      id="title"
    >
      Customers
    </h1>
    <a
      href="/customers/create"
    >
      <button
        id="button"
        type="button"
        value=""
      >
        <img
          alt="New"
          id="floppy"
          src="test-file-stub"
        />
        New Customer
      </button>
    </a>
  </div>
  <div>
    Error:
  </div>
</div>
`;

exports[`renders CustomersTable 1`] = `
<div
  class="content"
>
  <div
    class="top-panel"
  >
    <h1
      id="title"
    >
      Customers
    </h1>
    <a
      href="/customers/create"
    >
      <button
        id="button"
        type="button"
        value=""
      >
        <img
          alt="New"
          id="floppy"
          src="test-file-stub"
        />
        New Customer
      </button>
    </a>
  </div>
  <img
    alt="Loading...."
    id="load-img"
    src="test-file-stub"
  />
</div>
`;

exports[`renders CustomersTable 2`] = `
<div
  class="content"
>
  <div
    class="top-panel"
  >
    <h1
      id="title"
    >
      Customers
    </h1>
    <a
      href="/customers/create"
    >
      <button
        id="button"
        type="button"
        value=""
      >
        <img
          alt="New"
          id="floppy"
          src="test-file-stub"
        />
        New Customer
      </button>
    </a>
  </div>
  <table
    class="inv-table"
    id="inv-table"
  >
    <thead>
      <tr>
        <th>
          NAME
        </th>
        <th>
          PHONE
        </th>
        <th>
          EMAIL
        </th>
        <th>
          CREATED ON
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="newTD firstTD"
        >
          Jmp
        </td>
        <td
          class="newTD middleTD"
        />
        <td
          class="newTD middleTD"
        >
          J@rzp.com
        </td>
        <td
          class="newTD lastTD"
        >
          NaN Invalid Date NaN
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
